src/rate_limiter/src/lib.rs

```diff
@@ -158,22 +159,54 @@ impl TokenBucket {
+            self.last_update = now;
+        } else {
+            // At each 'time_delta' nanoseconds the bucket should refill with:
+            // refill_amount = (time_delta * size) / (complete_refill_time_ms * 1_000_000)
+            // `processed_capacity` and `processed_refill_time` are the result of simplifying above
+            // fraction formula with their greatest-common-factor.
+
+            // In the constructor, we assured that (self.refill_time * NANOSEC_IN_ONE_MILLISEC)
+            // fits into a u64 That means, at this point we know that time_delta <
+            // u64::MAX. Since all other values here are u64, this assures that u128
+            // multiplication cannot overflow.
+            let processed_capacity = u128::from(self.processed_capacity);
+            let processed_refill_time = u128::from(self.processed_refill_time);
+
+            let tokens = (time_delta * processed_capacity) / processed_refill_time;
+
+            // We increment `self.last_update` by the minimum time required to generate `tokens`, in
+            // the case where we have the time to generate `1.8` tokens but only
+            // generate `x` tokens due to integer arithmetic this will carry the time
+            // required to generate 0.8th of a token over to the next call, such that if
+            // the next call where to generate `2.3` tokens it would instead
+            // generate `3.1` tokens. This minimizes dropping tokens at high frequencies.
+            // We want the integer division here to round up instead of down (as if we round down,
+            // we would allow some fraction of a nano second to be used twice, allowing
+            // for the generation of one extra token in extreme circumstances).
+            let mut time_adjustment = tokens * processed_refill_time / processed_capacity;
+            if tokens * processed_refill_time % processed_capacity != 0 {
+                time_adjustment += 1;
+            }
 
-        // We increment `self.last_update` by the minimum time required to generate `tokens`, in the
-        // case where we have the time to generate `1.8` tokens but only generate `x` tokens due to
-        // integer arithmetic this will carry the time required to generate 0.8th of a token over to
-        // the next call, such that if the next call where to generate `2.3` tokens it would instead
-        // generate `3.1` tokens. This minimizes dropping tokens at high frequencies.
-        self.last_update +=
-            Duration::from_nanos((tokens * self.processed_refill_time) / self.processed_capacity);
-        self.budget = std::cmp::min(self.budget + tokens, self.size);
+            // Ensure that we always generate as many tokens as we can: assert that the "unused"
+            // part of time_delta is less than the time it would take to generate a
+            // single token (= processed_refill_time / processed_capacity)
+            debug_assert!(time_adjustment <= time_delta);
```

Comment:
Just a thought, but wondering if it would be worth adding a warning log line for release binaries in these cases.

Reply:
Mh, I like the idea of having non-fatal debug asserts to instead compile to warnings in release mode. Maybe part of @JonathanWoollett-Light's observability work?